File: src/main/java/domain/signal/TradingSignal.java
package domain.signal;

import java.time.Instant;


//===========================================================//
/**
 * Represents formatted trading signal that can be displayed
 */
//===========================================================//

public record TradingSignal(
  String symbol,
  Action action,
  Strength strength,
  double currentPrice,
  Long amount,
  Long currentStockCount,
  String reason,
  Instant createdAt
) {
   //===========================================================//
   //===========================================================//
   // Public Interface(s)

   public String formatToReadableText() {
      final String amountText = amount() == null
                                  ? "" : " | Amount:  " + amount();

      return symbol()
               + ": "
               + action()
               + " | "
               + strength()
               + " | Price: "
               + String.format("%.2f", currentPrice())
               + amountText
               + " | Current Stock Count: "
               + currentStockCount()
               + " | Reason: "
               + reason()
               + " | At: "
               + createdAt();
   }

   //===========================================================//
   //===========================================================//
   // Enum(s)

   /**
    * Represents the calculated Signal strength of generated signal.
    */
   public enum Strength {
      HIGH,
      MEDIUM,
      LOW
   }

   //===========================================================//
   /**
    * Represents the action suggested by the signal engine.
    */

   public enum Action {
      BUY,
      SELL,
      HOLD
   }
}

File: src/main/kotlin/domain/signal/TradingSignal.kt
package domain.signal

import domain.algorithm.TradingAlgorithm
import java.time.Instant

//===========================================================//
/**
 * Represents formatted trading signal that can be displayed
 */
//===========================================================//

class TradingSignal private constructor(
    val buy: TradingAlgorithm.Output.Buy?,
    val sell: TradingAlgorithm.Output.Sell?,
    val currentPrice: Double,
    val createdAt: Instant = Instant.now()
) {
    //===========================================================//
    //===========================================================//
    // Public Method(es)

    fun toReadableText(): String {
        val action = if(buy == null && sell == null) "HOLD"
        else if(buy != null && sell != null) "BUY, SELL"
        else if(buy != null) "BUY" else "SELL"

        val amount = if(buy == null && sell == null) ""
        else if(buy != null && sell != null) " | Buy Amount: ${buy.amount} | Sell: ${sell.batches.map { it.first }.toList()}"
        else if(buy != null) " | Buy Amount: ${buy.amount}" else " | Sell: ${sell!!.batches.map { it.first }.toList()}}"

        return ("" +
                action
                + " | Price: "
                + String.format("%.2f", currentPrice)
                + amount
                + " | At: "
                + createdAt)
    }

    //===========================================================//
    //===========================================================//
    // Constructor(s)

    constructor(output: TradingAlgorithm.Output, currentPrice: Double) : this(
        buy = output.buy,
        sell = output.sell,
        currentPrice = currentPrice
    )
}
